---
title: "`miro`: Visuals for Spatial (Transcript)Omics"
author:
- name: Helena L. Crowell
  affiliation: National Center for Genomics Analysis (CNAG), Barcelona, Spain
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 2
    toc_float: true
vignette: |
  %\VignetteIndexEntry{miro}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
date: "`r format(Sys.Date(), '%B %d, %Y')`"
package: "`r BiocStyle::pkg_ver('miro')`"
bibliography: refs.bib
---

# Preamble

## Introduction

`miro` provides a one-line command to visualize multi-layered data from
a .parquet-backed `Spatial-`, `SingleCellExperiment`, or `data.frame`.

To achieve this, data needs to be matched between objects with potentially 
different dimension names (e.g., spatial coordinates, cell identifiers). 

To try and avoid cumbersome passing of many arguments, `miro` tries to guess 
these from a catalog of values encountered across commercial platforms such 
as CosMx (NanoString) and Xenium (10x Genomics).

In addition, each layer is associated with a designed set of aesthetics, 
(e.g., color mapping, size etc.) allowing for highly flexible visualization.

## Dependencies

```{r load-libs, message=FALSE, warning=FALSE}
library(miro)
library(OSTA.data)
library(patchwork)
library(SpatialExperiment)
library(SpatialExperimentIO)
```

In this demo, we will rely on 10x Genomics Xenium data on a human CRC biopsy 
[@Oliveira2024], which comprises around around 400 RNA targets and 340k cells.

```{r load-data, message=FALSE, warning=FALSE}
# retrieve Xenium dataset
id <- "Xenium_HumanColon_Oliveira"
pa <- OSTA.data_load(id)
dir.create(td <- tempfile())
unzip(pa, exdir=td)
(spe <- readXeniumSXE(td))
```

# Visuals

## Centroids

By default (`pol=mol=NULL`), `miro` will visualize spatial locations as points;
in spatial transcriptomics data, these typically correspond to cell centroids.

The following parameter control this layer:

- `dat_xy`: spatial coordinates (defaults to `spatialCoordNames` for SPE)
- `dat_id`: observation identifier (e.g., cell ID)
- `dat_pal`: character vector of color values to use
- `dat_aes`: list of aesthetics passed to `geom_point()`
- `t`: transformation of values being colored by

```{r fig.width=4, fig.height=3.5}
miro(spe, dat_aes=list(stroke=0, size=0.2, col="total_counts"), t=log10)
```

For visualizations that include polygons and molecules, 
we specify two rather small regions of interest (ROIs):

```{r crop-data, message=FALSE, warning=FALSE}
# crops for demo
xy <- spatialCoords(spe)
spe$roi <- 
    xy[,1]>1000 & xy[,1]<2000 & 
    xy[,2]>1500 & xy[,2]<2500 
spe$roj <- 
    xy[,1]>1400 & xy[,1]<1600 & 
    xy[,2]>1800 & xy[,2]<2000 
```

Lastly, we specify which `metadata` slots contain molecules
and polygons (here, cell segmentation boundaries) of interest:

```{r pt}
p <- "cell_boundaries"
m <- "transcripts"
```

## Polygons

The following parameter control this layer:

- `pol_xy`: spatial coordinates (`miro` will try to guess)
- `pol_id`: observation identifier (e.g., cell ID)
- `pol_pal`: character vector of color values to use
- `pol_aes`: list of aesthetics passed to `geom_polygon()`
- `t`: transformation of values being colored by

```{r fig.height=4, fig.width=4.5}
.t <- \(.) ifelse(. > 300, 300, .) 
miro(spe, sub="roi", pol=p, pol_aes=list(fill="cell_area"), t=.t)
```

## Molecules

The following parameter control this layer:

- `mol_xy`: spatial coordinates (`miro` will try to guess)
- `mol_id`: observation identifier (e.g., cell ID)
- `mol_key`: feature identifier (e.g., `"symbol"`)
- `keys`: character vector of features to render
- `mol_pal`: character vector of color values to use
- `mol_aes`: list of aesthetics passed to `geom_point()`

```{r fig.height=4, fig.width=4.5}
miro(spe, sub="roi", 
    mol=m, mol_aes=list(size=0),
    keys=c("DUOX2", "MUC12", "VWF"))
```

# Layering

By default, centroids will be hidden when rendering `pol`ygons or `mol`ecules.
Setting `xy=TRUE` will force showing centroids regardless: 

```{r fig.width=3, fig.height=3}
miro(spe, sub="roj",
    pol=p, pol_aes=list(fill="darkgreen"),
    xy=TRUE, dat_aes=list(col="green", size=0.2))
```

Putting it all together, we can visualize centroids, polygons and molecules,
with separate control over each layer's aesthetics:

```{r fig.height=4, fig.width=4.5}
miro(spe, sub="roj",
    xy=TRUE, dat_aes=list(size=0.2, col="grey70"),
    pol=p, pol_aes=list(fill=NA, col="grey30", size=0.1),
    mol=m, mol_aes=list(size=0.1), 
    keys=c("DUOX2", "VWF", "MZB1"),
    mol_pal=c("magenta", "blue", "gold"))
```

## Subsetting

Argument `sub` may be used to restrict visuals to a subset of observation, and
accepts either an (integer) index or logical vector; when operating on SC/PE,
we can specify a pre-defined `colData` slot that contains a logical vector,
as already seen in the examples above:

```{r fig.width=7, fig.height=3.5}
aes <- list(fill="total_counts")
miro(spe, pol=p, sub="roi", pol_aes=aes) +
miro(spe, pol=p, sub="roj", pol_aes=aes) 
```

## Highlighting

Alternatively, we can use `hl` to highlight a subset of observations but not 
removing any; again, an index/logical vector or data slot may be used for this.
Color values of non-highlighted observations will be set to NA, such that `na`
may be used to control the color of these observations:

```{r fig.width=7, fig.height=3.5}
# subset 1st, highlight 2nd region
miro(spe, pol=p, sub="roi", hl="roj") +
# subset 2nd region
miro(spe, pol=p, sub="roj", hl=spe$cell_area < 100, na="gold")
```

# Appendix

## Session info {- .smaller}

<style type="text/css">
.smaller {
  font-size: 10px
}
</style>

```{r si, echo=FALSE}
sessionInfo()
```
